subsys/bluetooth/host/cs.c

```diff
@@ -360,72 +543,184 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+	}
+
+	p_result->header.procedure_counter = sys_le16_to_cpu(evt->procedure_counter);
+	p_result->header.frequency_compensation = sys_le16_to_cpu(evt->frequency_compensation);
+	p_result->header.procedure_done_status = evt->procedure_done_status;
+	p_result->header.subevent_done_status = evt->subevent_done_status;
+	p_result->header.procedure_abort_reason = evt->procedure_abort_reason;
+	p_result->header.subevent_abort_reason = evt->subevent_abort_reason;
+	p_result->header.reference_power_level = evt->reference_power_level;
+	p_result->header.num_antenna_paths = evt->num_antenna_paths;
+	p_result->header.num_steps_reported = evt->num_steps_reported;
+
+	p_result->header.config_id = 0;
+	p_result->header.start_acl_conn_event = 0;
+	if (conn) {
+		p_result->header.config_id = evt->config_id;
+		p_result->header.start_acl_conn_event =
+			sys_le16_to_cpu(evt->start_acl_conn_event_counter);
 	}
 
-#if defined(CONFIG_BT_CHANNEL_SOUNDING_TEST)
-	if (sys_le16_to_cpu(evt->conn_handle) == BT_HCI_LE_CS_TEST_CONN_HANDLE) {
-		result.header.config_id = 0;
-		result.header.start_acl_conn_event = 0;
+	if (evt->subevent_done_status != BT_HCI_LE_CS_SUBEVENT_DONE_STATUS_PARTIAL) {
+		invoke_subevent_result_callback(conn, p_result);
+	}
+
+	if (evt->procedure_done_status != BT_CONN_LE_CS_PROCEDURE_INCOMPLETE) {
+		/* We can now clear the any reassembly buffer allocated for this procedure,
+		 * to avoid code duplication, we're using the abort label to do so
+		 */
+		goto abort;
+	}
+
+	if (conn) {
+		bt_conn_unref(conn);
+		conn = NULL;
+	}
 
-		cs_test_callbacks.le_cs_test_subevent_data_available(&result);
+	return;
+
+abort:
+	if (conn) {
+		bt_conn_unref(conn);
+		conn = NULL;
+	}
+
+	reassembly_buf = get_reassembly_buf(conn_handle, false);
+	if (reassembly_buf) {
+		free_reassembly_buf(&reassembly_buf);
+	}
+}
+
+void bt_hci_le_cs_subevent_result_continue(struct net_buf *buf)
+{
+	struct bt_conn *conn = NULL;
+	struct bt_hci_evt_le_cs_subevent_result_continue *evt;
+	struct net_buf *reassembly_buf = NULL;
+	uint16_t conn_handle;
 
+	if (buf->len < sizeof(*evt)) {
+		LOG_ERR("Unexpected end of buffer");
+		return;
+	}
+
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+	conn_handle = sys_le16_to_cpu(evt->conn_handle);
+
+#if defined(CONFIG_BT_CHANNEL_SOUNDING_TEST)
+	if (conn_handle == BT_HCI_LE_CS_TEST_CONN_HANDLE) {
+		if (!cs_test_callbacks.le_cs_test_subevent_data_available) {
+			LOG_WRN("No callback registered. Discarded subevent results from CS Test.");
+			return;
+		}
 	} else
 #endif /* CONFIG_BT_CHANNEL_SOUNDING_TEST */
 	{
-		result.header.config_id = evt->config_id;
-		result.header.start_acl_conn_event =
-			sys_le16_to_cpu(evt->start_acl_conn_event_counter);
+		conn = bt_conn_lookup_handle(conn_handle, BT_CONN_TYPE_LE);
+		if (!conn) {
+			LOG_ERR("Unknown connection handle when processing subevent results");
+			return;
+		}
+	}
 
-		notify_cs_subevent_result(conn, &result);
+	uint16_t step_data_len = evt->num_steps_reported ? buf->len : 0;
```

Comment:
Not sure if num_steps_reported == 0 makes sense in a continue event. The spec seems to contradict itself a bit

`The Num_Steps_Reported parameter indicates the number of CS steps for which
results are reported. The Step_Mode[i] parameter indicates the CS mode for each CS
step, and the Step_Channel[i] parameter indicates the channel used for each CS step.
A Controller may return a value 0 when a subevent is aborted.`

but then

`For each CS procedure in a repeat sequence, the Controller shall report the results
for each CS step in each CS subevent of the CS procedure even if the CS step has been aborted for any reason.`

Reply:
The wording for subevent result event is:

> The HCI_LE_CS_Subevent_Result event shall be generated when the local Controller has results to report for a CS subevent during the CS procedure.

To me, this suggests that a controller doesn't have to wait until a subevent is finished to start reporting the results. Such a controller can send some partial results (whilst the procedure is running) but then send a continuation event to report that subevent/procedure was aborted with 0 remaining steps.